### ISSUE TITLE
Does this genuinely require OCaml 5.1?

### ISSUE THREAD

**Issue by aantron** [melange-community/melange-webapi#14]
I haven't looked into it in detail, but I suspect this constraint is too aggressive. It is presently causing a minor inconvenience in Dream's examples and CI.

**Comment by jchavarri** [comment]
Unfortunately Melange has its own [fork of OCaml compiler libs](https://github.com/melange-re/melange-compiler-libs), and it's not trivial to support multiple versions of the OCaml compiler without breaking tooling like [odoc](https://github.com/melange-re/melange/issues/603) or merlin/lsp, or adding a lot of maintenance overhead (there would need to be one version of Melange for each version of OCaml, e.g. `melange-2.0-ocaml-4.14`, `melange-2.0-ocaml-5.0`, `melange-2.0-ocaml-5.1`...

So Melange v2 is compatible with just [OCaml 5.1](https://github.com/melange-re/melange/blob/28b9e83d3c36998370185714589f182740da7e0f/melange.opam#L11). Most probably as we release new versions of Melange the range of OCaml versions supported will increase.

I am not sure what the fix can be. Probably have the Dream examples that depend on Melange only on 5.1...

 cc @anmonteiro

**Comment by jchavarri** [comment]
Btw, on most recent releases of Melange libraries, I am leaving the dep on `ocaml` "float" on the library opam (no constraint), so that the constraint on `ocaml` is defined only on `melange` itself, see https://github.com/ocaml/opam-repository/pull/24643#discussion_r1366514968.

**Comment by aantron** [comment]
Yes. If the library doesn't use specific features of OCaml 5.1, it should have a relaxed constraint. It should be possible to use the library with the lowest versions of OCaml and Melange that is actually required by the code of the library itself, and let the package manager figure out what versions of these deps to install and what the constraints between them are. Unless therre is something in melange-webapi 5.1 that absolutely requires the latest OCaml, this constraint is too aggressive and prevents useful installations of the library that would have worked.

**Comment by jchavarri** [comment]
I don't think I was clear.

The version of `melange-webapi` does not use anything from OCaml 5.1, but it is only compatible with Melange v2, which included breaking changes on the way bindings are defined.

So removing the `ocaml` constraint on `melange-webapi` will not change anything really. The library is not usable from OCaml 5.0 or earlier versions, because Melange v2 is only compatible with OCaml 5.1.

**Comment by jchavarri** [comment]
I think this issue should be opened on Melange repository, not here.

**Comment by aantron** [comment]
I did understand that and looked at the change, thank you! The constraint here is only redundant and misleading, and I think there is no actual issue, netiher here nor in Melange :) Thanks again!

**Comment by jchavarri** [comment]
Just curious, how did you work around the issues in CI you were facing in Dream?

Fwiw, I removed the `ocaml` constraint in https://github.com/melange-community/melange-webapi/pull/15, so the next released version of `melange-webapi` in the opam repository won't include it.

Regarding the generic issue, it would be great if Melange could support a broader range of versions of the underlying OCaml compiler while supporting packages like odoc and merlin/lsp. Unfortunately, I am not sure we have figured out a way to do this though. I started an issue in https://github.com/melange-re/melange/issues/886.
